cpp/oneapi/dal/detail/communicator.cpp

```diff
@@ -84,54 +75,15 @@ spmd::request_iface* spmd_communicator_via_host_impl::allgatherv(
 
     preview::detail::check_if_pointer_matches_queue(q, send_buf);
     preview::detail::check_if_pointer_matches_queue(q, recv_buf);
-    sycl::event::wait_and_throw(deps);
-
-    const std::int64_t rank_count = get_rank_count();
-    std::int64_t total_recv_count = 0;
 
-    array<std::int64_t> displs_host_root;
-    displs_host_root.reset(rank_count);
     {
-        std::int64_t* displs_host_root_ptr = displs_host_root.get_mutable_data();
-        for (std::int64_t i = 0; i < rank_count; i++) {
-            displs_host_root_ptr[i] = total_recv_count;
-            total_recv_count += recv_counts_host[i];
-        }
-    }
-
-    const std::int64_t dtype_size = get_data_type_size(dtype);
-    const std::int64_t send_size = check_mul_overflow(dtype_size, send_count);
-    const std::int64_t total_recv_size = check_mul_overflow(dtype_size, total_recv_count);
-    // Workaround for zero send_size
-    const auto send_buff_host = array<byte_t>::empty(send_size > 0 ? send_size : 1);
-    if (send_size > 0) {
-        memcpy_usm2host(q, send_buff_host.get_mutable_data(), send_buf, send_size);
-    }
-
-    array<byte_t> recv_buf_host;
-    byte_t* recv_buf_host_ptr = nullptr;
-    ONEDAL_ASSERT(total_recv_size > 0);
-    recv_buf_host.reset(total_recv_size);
-    recv_buf_host_ptr = recv_buf_host.get_mutable_data();
-    wait_request(allgatherv(send_buff_host.get_data(),
-                            send_count,
-                            recv_buf_host_ptr,
-                            recv_counts_host,
-                            displs_host,
-                            dtype));
-
-    const std::int64_t* displs_host_root_ptr = displs_host_root.get_data();
-    ONEDAL_ASSERT(displs_host_root_ptr);
-    ONEDAL_ASSERT(displs_host);
-    ONEDAL_ASSERT(recv_counts_host);
-
-    for (std::int64_t i = 0; i < rank_count; i++) {
-        const std::int64_t src_offset = check_mul_overflow(dtype_size, displs_host_root_ptr[i]);
-        const std::int64_t dst_offset = check_mul_overflow(dtype_size, displs_host[i]);
-        const std::int64_t copy_size = check_mul_overflow(dtype_size, recv_counts_host[i]);
-        if (copy_size > 0) {
-            memcpy_host2usm(q, recv_buf + dst_offset, recv_buf_host_ptr + src_offset, copy_size);
-        }
+        ONEDAL_PROFILER_TASK(comm.allgatherv_gpu, q);
```

Comment:
for allgatherv maybe it's necessary to add check_mul_overglow(num_ranks*one_rank_memory_size)

Reply:
actually looks like recv_counts_host is already allreduced counts - this function is dependent on a previous allreduce?

Reply:
should be resolved now - some unnecessary logic before conditional in case of real mpi but generally should work